### PR TITLE
`azurerm_kubernetes_cluster` - `api_server_vnet_integration_enabled` and `api_server_subnet_id` added (PREVIEW)

### LINKED ISSUE #18500: Support for AKS cluster with API Server VNet Integration
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

API Server VNET Integration for AKS cluster is in preview for both private and public clusters. It would be sweet to be able to provision it through Terraform. There are 2 modes for this feature:

- Managed VNet: when no VNet is provided, AKS will create the VNET and multiple subnets including one for the API server and configure them.
- BYO VNet: subnet for the API server needs to be created first and delegated to `Microsoft.ContainerService/managedClusters` (not sure about the corresponding `actions` in my example though) and passed to the AKS cluster. We need also that the AKS cluster Managed Identity to have `Network Contributor` roles on both AKS cluster and API server subnets. 

### New or Affected Resource(s)/Data Source(s)

azurerm_kubernetes_cluster

### Potential Terraform Configuration

```hcl
resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}

resource "azurerm_virtual_network" "example" {
  name                = "example-vnet"
  address_space       = ["10.0.0.0/16"]
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name
}

resource "azurerm_subnet" "node_subnet" {
  name                 = "node-subnet"
  resource_group_name  = azurerm_resource_group.example.name
  virtual_network_name = azurerm_virtual_network.example.name
  address_prefixes     = ["10.0.1.0/24"]
}

resource "azurerm_subnet" "apiserver_subnet" {
  name                 = "apiserver-subnet"
  resource_group_name  = azurerm_resource_group.example.name
  virtual_network_name = azurerm_virtual_network.example.name
  address_prefixes     = ["10.0.2.0/28"]

  delegation {
    name = "delegation"

    service_delegation {
      name    = "Microsoft.ContainerService/managedClusters"
      actions = ["Microsoft.Network/virtualNetworks/subnets/join/action", "Microsoft.Network/virtualNetworks/subnets/prepareNetworkPolicies/action"]
    }
  }
}

resource "azurerm_user_assigned_identity" "example" {
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location

  name = "aks"
}

resource "azurerm_role_assignment" "aks_subnet_role" {
  scope                = azurerm_subnet.node_subnet.id
  role_definition_name = "Network Contributor""
  principal_id         = azurerm_user_assigned_identity.example.id
}

resource "azurerm_role_assignment" "apiserver_subnet_role" {
  scope                = azurerm_subnet.apiserver_subnet.id
  role_definition_name = "Network Contributor""
  principal_id         = azurerm_user_assigned_identity.example.id
}

resource "azurerm_kubernetes_cluster" "example" {
  name                               = "example-aks1"
  location                           = azurerm_resource_group.example.location
  resource_group_name                = azurerm_resource_group.example.name
  dns_prefix                         = "exampleaks1"
  apiserver_vnet_integration_enabled = true
  apiserver_subnet_id                = azurerm_subnet.apiserver_subnet.id

  default_node_pool {
    name       = "default"
    node_count = 1
    vm_size    = "Standard_D2_v2"
  }

  identity {
    type = "SystemAssigned"
  }

  tags = {
    Environment = "Production"
  }

  depends_on = [
    azurerm_role_assignment.aks_subnet_role,
    azurerm_role_assignment.apiserver_subnet_role
  ]
}
```

### References

https://learn.microsoft.com/en-us/azure/aks/api-server-vnet-integration

### PR DESCRIPTION
**Pull request by aristosvo**
Fixes #18500

Needs `az feature register --namespace "Microsoft.ContainerService" --name "EnableAPIServerVnetIntegrationPreview"` to work

- [x] Managed and BYO VNet Acctest
- [x] Docs
- [x] All use-cases and validation scenario's tested/documented/implemented

### CODE REVIEW

File: internal/services/containers/kubernetes_cluster_resource.go [hunk 1]
```diff
@@ -99,6 +99,18 @@ func resourceKubernetesCluster() *pluginsdk.Resource {
...
+				Type:     pluginsdk.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"api_server_subnet_id": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ValidateFunc: azure.ValidateResourceID,
+			},
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_resource.go:112]: Wondering whether it makes sense to move all the `api_server_` properties into a block called `api_server_profile` and deprecate and remove `api_server_authorized_ip_ranges` for 4.0?

File: internal/services/containers/kubernetes_cluster_resource.go [hunk 2]
```diff
@@ -99,6 +99,18 @@ func resourceKubernetesCluster() *pluginsdk.Resource {
...
+			"api_server_vnet_integration_enabled": {
+				Type:     pluginsdk.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"api_server_subnet_id": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ValidateFunc: azure.ValidateResourceID,
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_resource.go:111]: We should be able to use the validation function for the subnet Id here
```suggestion
				ValidateFunc: networkValidate.SubnetID,
```

File: internal/services/containers/kubernetes_cluster_resource.go [hunk 3]
```diff
@@ -1928,6 +1958,12 @@ func resourceKubernetesClusterRead(d *pluginsdk.ResourceData, meta interface{})
 			if err := d.Set("api_server_authorized_ip_ranges", apiServerAuthorizedIPRanges); err != nil {
 				return fmt.Errorf("setting `api_server_authorized_ip_ranges`: %+v", err)
 			}
+			if accessProfile.EnableVnetIntegration != nil {
+				enableAPIServerVnetIntegration = *accessProfile.EnableVnetIntegration
+			}
+			if accessProfile.SubnetId != nil && *accessProfile.SubnetId != "" {
+				d.Set("api_server_subnet_id", accessProfile.SubnetId)
+			}
```
Comment (stephybun) [inline internal/services/containers/kubernetes_cluster_resource.go:1966]: This should always have a value set into state so can this become 
```suggestion
			apiServerSubnetId := ""
			if accessProfile.SubnetId != nil {
				apiServerSubnetId = *accessProfile.SubnetId
			}
			d.Set("api_server_subnet_id", apiServerSubnetId)
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for this PR @aristosvo. This looks good to me, I left a few suggestions/questions in-line. Once those are resolved we can take another look through and get this merged!

File: internal/services/containers/kubernetes_cluster_resource.go (hunk 1 above)
Comment (aristosvo) [reply internal/services/containers/kubernetes_cluster_resource.go:115]: @stephybun Follow up question: would you advise refactoring all `ManagedClusterAPIServerAccessProfile` properties or only the properties that are named explicitly `api_server_*`?

```go
apiAccessProfile := managedclusters.ManagedClusterAPIServerAccessProfile{
		EnablePrivateCluster:           &enablePrivateCluster,
		AuthorizedIPRanges:             apiServerAuthorizedIPRanges,
		EnablePrivateClusterPublicFQDN: utils.Bool(d.Get("private_cluster_public_fqdn_enabled").(bool)),
		DisableRunCommand:              utils.Bool(!d.Get("run_command_enabled").(bool)),
		EnableVnetIntegration:          &enableAPIServerVnetIntegration,
		SubnetId:                       utils.String(d.Get("api_server_subnet_id").(string)),
	}
```

**Comment by aristosvo** [comment]
Tests are passing after the last modifications:
```bash
❯ go install && make acctests SERVICE='containers' TESTARGS='-run=TestAccKubernetesCluster_apiServerIn'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/containers -run=TestAccKubernetesCluster_apiServerIn -timeout 180m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccKubernetesCluster_apiServerInManagedSubnet
=== PAUSE TestAccKubernetesCluster_apiServerInManagedSubnet
=== RUN   TestAccKubernetesCluster_apiServerInBYOSubnet
=== PAUSE TestAccKubernetesCluster_apiServerInBYOSubnet
=== CONT  TestAccKubernetesCluster_apiServerInManagedSubnet
=== CONT  TestAccKubernetesCluster_apiServerInBYOSubnet
--- PASS: TestAccKubernetesCluster_apiServerInManagedSubnet (776.48s)
--- PASS: TestAccKubernetesCluster_apiServerInBYOSubnet (1590.64s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/containers   1592.398s
```

File: internal/services/containers/kubernetes_cluster_resource.go [hunk 4]
```diff
@@ -2077,8 +2095,9 @@ func resourceKubernetesClusterRead(d *pluginsdk.ResourceData, meta interface{})
 				d.Set("private_dns_zone_id", accessProfile.PrivateDNSZone)
 			}
 		}
-
-		d.Set("private_cluster_enabled", enablePrivateCluster)
+		if !features.FourPointOhBeta() {
+			d.Set("private_cluster_enabled", enablePrivateCluster)
+		}
```
[the remark this answers is not in the document]
Comment (aristosvo) [reply internal/services/containers/kubernetes_cluster_resource.go]: Thanks!

**Review by stephybun — APPROVED** [review]
Thanks @aristosvo LGTM 🌮

**Pull request merged**